File: examples/typescript/psea-verify.ts
#!/usr/bin/env node
/**
 * PSEA reference verifier — TypeScript / Node 22+.
 *
 * Verify-only. Implements the algorithms in `spec/tier-definitions.md`.
 * Walks ../../test-vectors/ and reports PASS/FAIL per vector.
 *
 * Zero npm dependencies — uses only Node's built-in `crypto` and `fs`.
 *
 * Run:
 *   node --experimental-strip-types psea-verify.ts
 *   # or
 *   npm run verify
 */

import { readFileSync, readdirSync, statSync } from "node:fs";
import { join, resolve, relative } from "node:path";
import { createHash, createPublicKey, createVerify } from "node:crypto";
import { fileURLToPath } from "node:url";

const HERE = fileURLToPath(new URL(".", import.meta.url));
const ROOT = resolve(HERE, "..", "..");
const VECTORS = join(ROOT, "test-vectors");

const NOW = 1_733_280_000_000;
const MAX_CLOCK_SKEW_MS = 30_000;
const BIOMETRIC_FRESHNESS_MS = 30_000;

// ---------------------------------------------------------------------------
// Helpers

function jcs(value: unknown): string {
  // RFC 8785 minimal: sort keys recursively, no whitespace.
  if (value === null || typeof value !== "object") {
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    return "[" + value.map(jcs).join(",") + "]";
  }
  const keys = Object.keys(value as Record<string, unknown>).sort();
  return (
    "{" +
    keys
      .map(
        (k) => JSON.stringify(k) + ":" + jcs((value as Record<string, unknown>)[k])
      )
      .join(",") +
    "}"
  );
}

function jcsBytes(value: unknown): Buffer {
  return Buffer.from(jcs(value), "utf8");
}

function sha256Hex(b: Buffer): string {
  return createHash("sha256").update(b).digest("hex");
}

function b64uDecode(s: string): Buffer {
  const pad = "=".repeat((4 - (s.length % 4)) % 4);
  return Buffer.from(s.replace(/-/g, "+").replace(/_/g, "/") + pad, "base64");
}

function spkiFromUncompressedP256(pubB64u: string): import("node:crypto").KeyObject {
  const raw = b64uDecode(pubB64u);
  if (raw.length !== 65 || raw[0] !== 0x04) {
    throw new Error("expected 65-byte uncompressed P-256 public key");
  }
  // Build SubjectPublicKeyInfo for P-256 (NIST secp256r1)
  // SEQ(SEQ(OID ec-pub-key, OID prime256v1), BIT STRING raw)
  const algId = Buffer.from(
    "3013" +                       // SEQUENCE 0x13
    "06072a8648ce3d0201" +         // OID 1.2.840.10045.2.1 (ecPublicKey)
    "06082a8648ce3d030107",        // OID 1.2.840.10045.3.1.7 (prime256v1)
    "hex",
  );
  const bitString = Buffer.concat([Buffer.from([0x03, raw.length + 1, 0x00]), raw]);
  const inner = Buffer.concat([algId, bitString]);
  const spki = Buffer.concat([Buffer.from([0x30, 0x82, (inner.length >> 8) & 0xff, inner.length & 0xff]), inner]);
  return createPublicKey({ key: spki, format: "der", type: "spki" });
}

function rawSigToDer(raw: Buffer): Buffer {
  if (raw.length !== 64) throw new Error("expected 64-byte raw signature");
  const r = trimLeadingZeros(raw.subarray(0, 32));
  const s = trimLeadingZeros(raw.subarray(32, 64));
  const rEnc = encodeInt(r);
  const sEnc = encodeInt(s);
  const inner = Buffer.concat([rEnc, sEnc]);
  return Buffer.concat([Buffer.from([0x30, inner.length]), inner]);
}

function trimLeadingZeros(b: Buffer): Buffer {
  let i = 0;
  while (i < b.length - 1 && b[i] === 0) i++;
  return b.subarray(i);
}

function encodeInt(b: Buffer): Buffer {
  // Add a leading 0x00 if high bit set.
  const needsPad = (b[0] & 0x80) !== 0;
  const body = needsPad ? Buffer.concat([Buffer.from([0x00]), b]) : b;
  return Buffer.concat([Buffer.from([0x02, body.length]), body]);
}

function verifyES256(pub: import("node:crypto").KeyObject, msg: Buffer, rawSig: Buffer): boolean {
  try {
    const der = rawSigToDer(rawSig);
    const v = createVerify("sha256");
    v.update(msg);
    v.end();
    return v.verify(pub, der);
  } catch {
    return false;
  }
}

// ---------------------------------------------------------------------------
// Registry

type RegistryEntry = {
  publicKey: string;
  trustState: string;
  trustStateHash: string;
  counter: number;
  chainHead: string;
  attestationNotAfter: number;
  tenantPolicy: string;
};

class Registry {
  fixedNowMs: number;
  devices: Record<string, RegistryEntry>;
  pubkeys: Record<string, import("node:crypto").KeyObject>;

  constructor(raw: any) {
    this.fixedNowMs = raw.fixedTimestampMs as number;
    this.devices = raw.verifierRegistry as Record<string, RegistryEntry>;
    this.pubkeys = {};
    for (const [did, entry] of Object.entries(this.devices)) {
      this.pubkeys[did] = spkiFromUncompressedP256(entry.publicKey);
    }
  }

  static load(p: string): Registry {
    const raw = JSON.parse(readFileSync(p, "utf8"));
    return new Registry(raw);
  }
}

// ---------------------------------------------------------------------------
// Per-tier verifiers

const RISK_TIER_MIN: Record<string, string> = { LOW: "P", MEDIUM: "S", HIGH: "E", CRITICAL: "A" };
const TIER_RANK: Record<string, number> = { P: 0, S: 1, E: 2, A: 3 };

function insufficientTier(tier: string, actionClass: string): boolean {
  return TIER_RANK[tier] < TIER_RANK[RISK_TIER_MIN[actionClass]];
}

function verifyTierP(reg: Registry, vec: any): string {
  const body = vec.proofToken.body;
  if (insufficientTier("P", vec.actionClass)) return "REJECT_INSUFFICIENT_TIER";
  const did = body.deviceId;
  if (!(did in reg.devices)) return "REJECT_UNKNOWN_DEVICE";
  if (body.trustStateHash !== reg.devices[did].trustStateHash) return "REJECT_STATE_DRIFT";
  if (Math.abs(reg.fixedNowMs - body.timestamp) > MAX_CLOCK_SKEW_MS)
    return "REJECT_TIMESTAMP_OUT_OF_RANGE";
  return "ACCEPT";
}

function verifyTierS(reg: Registry, vec: any): string {
  const body = vec.proofToken.body;
  if (insufficientTier("S", vec.actionClass)) return "REJECT_INSUFFICIENT_TIER";

  const expectedHash = sha256Hex(jcsBytes(vec.input.actionRequest));
  if (body.actionHash !== expectedHash) return "REJECT_ACTION_BINDING";

  const sess = vec.input.deviceState.session;
  if (!sess || sess.id !== body.sessionId) return "REJECT_UNKNOWN_SESSION";

  const did = body.deviceId;
  const pub = reg.pubkeys[did];
  if (!pub) return "REJECT_UNKNOWN_DEVICE";

  const sig = b64uDecode(vec.proofToken.signature);
  if (!verifyES256(pub, jcsBytes(body), sig)) return "REJECT_BAD_SIGNATURE";
  return "ACCEPT";
}

function verifyTierECommon(reg: Registry, vec: any, allowTest = false): string {
  const body = vec.proofToken.body;
  const state = vec.input.deviceState;

  const expectedHash = sha256Hex(jcsBytes(vec.input.actionRequest));
  if (body.actionHash !== expectedHash) return "REJECT_ACTION_BINDING";

  const did = body.deviceId;
  const pub = reg.pubkeys[did];
  if (!pub) return "REJECT_UNKNOWN_DEVICE";

  const sig = b64uDecode(vec.proofToken.signature);
  if (!verifyES256(pub, jcsBytes(body), sig)) return "REJECT_BAD_SIGNATURE";

  if (body.counter <= state.counter) return "REJECT_REPLAY";

  const { chainEntry: _ignored, ...bodyNoChain } = body;
  const expectedEntry = sha256Hex(
    Buffer.concat([Buffer.from(state.chainHead, "hex"), jcsBytes(bodyNoChain)]),
  );
  if (body.chainEntry !== expectedEntry) return "REJECT_CHAIN_BROKEN";
  if (body.chainPrev !== state.chainHead) return "REJECT_CHAIN_BROKEN";

  const bio = body.biometric;
  if (!bio) return "REJECT_BIOMETRIC";
  const age = body.timestamp - bio.freshness.capturedAt;
  if (age < 0 || age > BIOMETRIC_FRESHNESS_MS) return "REJECT_BIOMETRIC";

  const att = body.attestation;
  if (att.source === "test" && !allowTest) return "REJECT_ATTESTATION";
  if (att.notAfter < reg.fixedNowMs) return "REJECT_ATTESTATION";
  if (att.packageBinding.enforcement === "client-claim" &&
      ["REGULATED", "STANDARD"].includes(state.tenantPolicy)) {
    return "REJECT_ATTESTATION";
  }
  return "ACCEPT";
}

function verifyTierE(reg: Registry, vec: any): string {
  if (insufficientTier("E", vec.actionClass)) return "REJECT_INSUFFICIENT_TIER";
  return verifyTierECommon(reg, vec);
}

function verifyTierA(reg: Registry, vec: any): string {
  if (insufficientTier("A", vec.actionClass)) return "REJECT_INSUFFICIENT_TIER";
  const state = vec.input.deviceState;
  if (!state.networkReachable || !state.authorizeEndpointHealthy) {
    return "REJECT_OFFLINE_AUTHORITATIVE_BLOCKED";
  }
  const inner = verifyTierECommon(reg, vec);
  if (inner !== "ACCEPT") return inner;

  const blocked: string[] = state.blockedActionTypes ?? [];
  if (blocked.includes(vec.input.actionRequest.actionType)) {
    return "REJECT_POLICY_DENY";
  }
  if (state.auditLogHealthy === false) return "REJECT_AUDIT_FAILURE";
  return "ACCEPT";
}

const VERIFIERS: Record<string, (r: Registry, v: any) => string> = {
  P: verifyTierP, S: verifyTierS, E: verifyTierE, A: verifyTierA,
};

// ---------------------------------------------------------------------------
// Runner

function listVectors(): string[] {
  const out: string[] = [];
  for (const tier of ["tier-p", "tier-s", "tier-e", "tier-a"]) {
    const dir = join(VECTORS, tier);
    if (!statSync(dir, { throwIfNoEntry: false })) continue;
    for (const f of readdirSync(dir).sort()) {
      if (f.endsWith(".json")) out.push(join(dir, f));
    }
  }
  return out;
}

function main(): number {
  const reg = Registry.load(join(VECTORS, "keys", "test-keys.json"));
  const files = listVectors();
  let pass = 0, fail = 0;

  for (const f of files) {
    const vec = JSON.parse(readFileSync(f, "utf8"));
    const tier = vec.proofToken.header.tier as string;
    const actual = VERIFIERS[tier](reg, vec);
    const expected = vec.expected.result as string;
    const rel = relative(ROOT, f).replaceAll("\\", "/");
    if (actual === expected) {
      pass++;
      console.log(`  [PASS] ${rel}  (${actual})`);
    } else {
      fail++;
      console.log(`  [FAIL] ${rel}  expected=${expected} actual=${actual}`);
    }
  }
  console.log(`\n${pass} passed, ${fail} failed, ${files.length} total`);
  return fail === 0 ? 0 : 1;
}

process.exit(main());

File: examples/python/psea_verify.py
#!/usr/bin/env python3
"""
PSEA reference verifier — Python.

Verify-only. Implements the algorithms in `spec/tier-definitions.md`.
Walks ../../test-vectors/ and reports PASS/FAIL per vector.

No private-key code. No signing. Reads only public keys from the test-keys
fixture.

Usage:
    pip install cryptography
    python3 psea_verify.py
"""

from __future__ import annotations

import hashlib
import json
import sys
from base64 import urlsafe_b64decode
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives.asymmetric.utils import (
    encode_dss_signature,
)

ROOT = Path(__file__).resolve().parents[2]
VECTORS = ROOT / "test-vectors"

NOW = 1_733_280_000_000  # match bootstrap-vectors.py
MAX_CLOCK_SKEW_MS = 30_000
BIOMETRIC_FRESHNESS_MS = 30_000


# ---------------------------------------------------------------------------
# Helpers


def jcs(obj: Any) -> bytes:
    """RFC 8785 JSON Canonicalization (minimal)."""
    return json.dumps(obj, sort_keys=True, separators=(",", ":"),
                      ensure_ascii=False).encode("utf-8")


def b64u_decode(s: str) -> bytes:
    pad = "=" * (-len(s) % 4)
    return urlsafe_b64decode(s + pad)


def sha256_hex(b: bytes) -> str:
    return hashlib.sha256(b).hexdigest()


def load_public_key(b64u_uncompressed: str) -> ec.EllipticCurvePublicKey:
    raw = b64u_decode(b64u_uncompressed)
    if len(raw) != 65 or raw[0] != 0x04:
        raise ValueError("expected 65-byte uncompressed P-256 public key")
    return ec.EllipticCurvePublicKey.from_encoded_point(ec.SECP256R1(), raw)


def verify_es256_raw(pub: ec.EllipticCurvePublicKey, msg: bytes,
                     raw_sig: bytes) -> bool:
    if len(raw_sig) != 64:
        return False
    r = int.from_bytes(raw_sig[:32], "big")
    s = int.from_bytes(raw_sig[32:], "big")
    der = encode_dss_signature(r, s)
    try:
        pub.verify(der, msg, ec.ECDSA(hashes.SHA256()))
        return True
    except InvalidSignature:
        return False


# ---------------------------------------------------------------------------
# Verifier registry (loaded from test-keys.json)


@dataclass
class Registry:
    keys: dict
    devices: dict          # deviceId -> dict from verifierRegistry
    pubkeys: dict          # deviceId -> EllipticCurvePublicKey
    fixed_now_ms: int

    @classmethod
    def load(cls, p: Path) -> "Registry":
        raw = json.loads(p.read_text(encoding="utf-8"))
        pubkeys = {}
        for did, entry in raw["verifierRegistry"].items():
            pubkeys[did] = load_public_key(entry["publicKey"])
        return cls(keys=raw, devices=raw["verifierRegistry"],
                   pubkeys=pubkeys, fixed_now_ms=raw["fixedTimestampMs"])


# ---------------------------------------------------------------------------
# Per-tier verifiers


RISK_TIER_MIN = {"LOW": "P", "MEDIUM": "S", "HIGH": "E", "CRITICAL": "A"}
TIER_RANK = {"P": 0, "S": 1, "E": 2, "A": 3}


def insufficient_tier(tier: str, action_class: str) -> bool:
    return TIER_RANK[tier] < TIER_RANK[RISK_TIER_MIN[action_class]]


def verify_tier_p(reg: Registry, vec: dict) -> str:
    body = vec["proofToken"]["body"]
    if insufficient_tier("P", vec["actionClass"]):
        return "REJECT_INSUFFICIENT_TIER"
    did = body["deviceId"]
    if did not in reg.devices:
        return "REJECT_UNKNOWN_DEVICE"
    if body["trustStateHash"] != reg.devices[did]["trustStateHash"]:
        return "REJECT_STATE_DRIFT"
    if abs(reg.fixed_now_ms - body["timestamp"]) > MAX_CLOCK_SKEW_MS:
        return "REJECT_TIMESTAMP_OUT_OF_RANGE"
    return "ACCEPT"


def verify_tier_s(reg: Registry, vec: dict) -> str:
    body = vec["proofToken"]["body"]
    if insufficient_tier("S", vec["actionClass"]):
        return "REJECT_INSUFFICIENT_TIER"

    expected_hash = sha256_hex(jcs(vec["input"]["actionRequest"]))
    if body["actionHash"] != expected_hash:
        return "REJECT_ACTION_BINDING"

    sess = vec["input"]["deviceState"].get("session")
    if not sess or sess.get("id") != body["sessionId"]:
        return "REJECT_UNKNOWN_SESSION"

    did = body["deviceId"]
    pub = reg.pubkeys.get(did)
    if pub is None:
        return "REJECT_UNKNOWN_DEVICE"

    sig = b64u_decode(vec["proofToken"]["signature"])
    if not verify_es256_raw(pub, jcs(body), sig):
        return "REJECT_BAD_SIGNATURE"

    return "ACCEPT"


def verify_tier_e_common(reg: Registry, vec: dict, *, allow_test: bool = False) -> str:
    body  = vec["proofToken"]["body"]
    state = vec["input"]["deviceState"]

    expected_hash = sha256_hex(jcs(vec["input"]["actionRequest"]))
    if body["actionHash"] != expected_hash:
        return "REJECT_ACTION_BINDING"

    did = body["deviceId"]
    pub = reg.pubkeys.get(did)
    if pub is None:
        return "REJECT_UNKNOWN_DEVICE"

    sig = b64u_decode(vec["proofToken"]["signature"])
    if not verify_es256_raw(pub, jcs(body), sig):
        return "REJECT_BAD_SIGNATURE"

    if body["counter"] <= state["counter"]:
        return "REJECT_REPLAY"

    body_no_chain = {k: v for k, v in body.items() if k != "chainEntry"}
    expected_entry = sha256_hex(bytes.fromhex(state["chainHead"]) + jcs(body_no_chain))
    if body["chainEntry"] != expected_entry:
        return "REJECT_CHAIN_BROKEN"
    if body["chainPrev"] != state["chainHead"]:
        return "REJECT_CHAIN_BROKEN"

    bio = body.get("biometric")
    if bio is None:
        return "REJECT_BIOMETRIC"
    age = body["timestamp"] - bio["freshness"]["capturedAt"]
    if age < 0 or age > BIOMETRIC_FRESHNESS_MS:
        return "REJECT_BIOMETRIC"

    att = body["attestation"]
    if att["source"] == "test" and not allow_test:
        return "REJECT_ATTESTATION"
    if att["notAfter"] < reg.fixed_now_ms:
        return "REJECT_ATTESTATION"
    if att["packageBinding"]["enforcement"] == "client-claim" \
            and state.get("tenantPolicy") in ("REGULATED", "STANDARD"):
        # Production policies forbid client-claim
        return "REJECT_ATTESTATION"

    return "ACCEPT"


def verify_tier_e(reg: Registry, vec: dict) -> str:
    if insufficient_tier("E", vec["actionClass"]):
        return "REJECT_INSUFFICIENT_TIER"
    return verify_tier_e_common(reg, vec)


def verify_tier_a(reg: Registry, vec: dict) -> str:
    if insufficient_tier("A", vec["actionClass"]):
        return "REJECT_INSUFFICIENT_TIER"

    state = vec["input"]["deviceState"]
    if not state.get("networkReachable") or not state.get("authorizeEndpointHealthy"):
        return "REJECT_OFFLINE_AUTHORITATIVE_BLOCKED"

    inner = verify_tier_e_common(reg, vec)
    if inner != "ACCEPT":
        return inner

    body = vec["proofToken"]["body"]
    blocked = state.get("blockedActionTypes", [])
    action_type = vec["input"]["actionRequest"].get("actionType")
    if action_type in blocked:
        return "REJECT_POLICY_DENY"

    if state.get("auditLogHealthy", True) is False:
        return "REJECT_AUDIT_FAILURE"

    return "ACCEPT"


VERIFIERS = {"P": verify_tier_p, "S": verify_tier_s,
             "E": verify_tier_e, "A": verify_tier_a}


# ---------------------------------------------------------------------------
# Test runner


def run_one(reg: Registry, vec_path: Path) -> tuple[bool, str, str]:
    vec  = json.loads(vec_path.read_text(encoding="utf-8"))
    tier = vec["proofToken"]["header"]["tier"]
    actual = VERIFIERS[tier](reg, vec)
    expected = vec["expected"]["result"]
    return (actual == expected, actual, expected)


def main() -> int:
    if not VECTORS.exists():
        print(f"FATAL: {VECTORS} missing — run tools/bootstrap-vectors.py first",
              file=sys.stderr)
        return 2
    reg = Registry.load(VECTORS / "keys" / "test-keys.json")

    files = sorted(VECTORS.glob("tier-*/*.json"))
    if not files:
        print("FATAL: no test vectors found", file=sys.stderr)
        return 2

    pass_count = fail_count = 0
    for f in files:
        ok, actual, expected = run_one(reg, f)
        marker = "PASS" if ok else "FAIL"
        rel = f.relative_to(ROOT).as_posix()
        if ok:
            pass_count += 1
            print(f"  [{marker}] {rel}  ({actual})")
        else:
            fail_count += 1
            print(f"  [{marker}] {rel}  expected={expected} actual={actual}")

    print(f"\n{pass_count} passed, {fail_count} failed, {len(files)} total")
    return 0 if fail_count == 0 else 1


if __name__ == "__main__":
    sys.exit(main())
